File: proto-csharp/ProtoMessage/ProtoMessage/ApeMessage/Reflect/ListFieldDescriptor.cs
using System;
using System.Reflection;
using System.Collections.Generic;
using Google.ProtocolBuffers;
using com.spaceape.protobuf;

namespace com.spaceape.protobuf.reflect
{
  public class ListFieldDescriptor<T> : FieldDescriptor
  {
    private FieldDescriptor repeatingDescriptor;

    public ListFieldDescriptor(int fieldNumber, String fieldName, Type clazz, FieldDescriptor repeatingDescriptor) : base(fieldNumber, fieldName, clazz)
    {
      this.repeatingDescriptor = repeatingDescriptor;
    }

    override public Object readField(CodedInputStream input, CodeGenContext context)
    {
      int size = input.ReadInt32();
      List<T> newlist = new List<T>(size);
      for (int i = 0; i < size; i++)
      {
        newlist.Add((T)repeatingDescriptor.readField(input, context));
      }
      return newlist;
    }

    override public void writeField(CodedOutputStream output, CodeGenContext context, Object obj)
    {
      List<T> list = (List<T>)obj;
      output.WriteInt32(fieldNumber, list.Count);
      for (int i = 0; i < list.Count; i++)
      {
        repeatingDescriptor.writeField(output, context, list[i]);
      }
    }
  }
}

File: proto-java/src/main/java/com/spaceape/protobuf/reflect/ListFieldDescriptor.java
package com.spaceape.protobuf.reflect;

import com.google.protobuf.CodedInputStream;
import com.google.protobuf.CodedOutputStream;
import com.spaceape.protobuf.CodeGenContext;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class ListFieldDescriptor<T> extends FieldDescriptor{
    public final FieldDescriptor repeatingDescriptor;

    public ListFieldDescriptor(int fieldNumber, String fieldName, Class clazz, FieldDescriptor repeatingDescriptor){
        super(fieldNumber, fieldName, clazz, List.class);
        this.repeatingDescriptor = repeatingDescriptor;
    }
    public List<T> readField(CodedInputStream input, CodeGenContext context) throws IOException{
        int size = input.readInt32();
        List<T> newlist = new ArrayList<T>(size);
        for (int i=0;i<size;i++){
            int tag = input.readTag();
            newlist.add((T)repeatingDescriptor.readField(input, context));
        }
        return newlist;
    }

    public void writeField(CodedOutputStream output, CodeGenContext context, Object object) throws IOException{
        List<T> list = (List<T>)object;
        output.writeInt32(fieldNumber, list.size());
        for (int i=0;i<list.size();i++){
            repeatingDescriptor.writeField(output, context, list.get(i));
        }
    }
}
